File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcCapabilityCatalog.cs
namespace Sdkwork.Rtc.Sdk;

using System.Collections.Generic;
using System.Linq;

public sealed record RtcCapabilityCatalogEntry(
    string capabilityKey,
    string category,
    string surface
);

public static class RtcCapabilityCatalog
{
    public static readonly IReadOnlyList<RtcCapabilityCatalogEntry> Entries =
    [
        new("session", "required-baseline", "cross-surface"),
        new("join", "required-baseline", "runtime-bridge"),
        new("publish", "required-baseline", "runtime-bridge"),
        new("subscribe", "required-baseline", "runtime-bridge"),
        new("mute", "required-baseline", "runtime-bridge"),
        new("basic-events", "required-baseline", "runtime-bridge"),
        new("health", "required-baseline", "control-plane"),
        new("unwrap", "required-baseline", "runtime-bridge"),
        new("screen-share", "optional-advanced", "runtime-bridge"),
        new("recording", "optional-advanced", "control-plane"),
        new("cloud-mix", "optional-advanced", "control-plane"),
        new("cdn-relay", "optional-advanced", "control-plane"),
        new("data-channel", "optional-advanced", "runtime-bridge"),
        new("transcription", "optional-advanced", "control-plane"),
        new("beauty", "optional-advanced", "runtime-bridge"),
        new("spatial-audio", "optional-advanced", "runtime-bridge"),
        new("e2ee", "optional-advanced", "runtime-bridge"),
    ];

public static IReadOnlyList<RtcCapabilityCatalogEntry> GetRtcCapabilityCatalog() =>
        Entries;

    public static RtcCapabilityCatalogEntry? GetRtcCapabilityDescriptor(string capabilityKey) =>
        Entries.FirstOrDefault(entry => entry.capabilityKey == capabilityKey);

}

File: sdkwork-rtc-sdk-flutter/lib/src/rtc_capability_catalog.dart
final class RtcCapabilityCatalogEntry {
  const RtcCapabilityCatalogEntry({
    required this.capabilityKey,
    required this.category,
    required this.surface,
  });

  final String capabilityKey;
  final String category;
  final String surface;
}

final class RtcCapabilityCatalog {
  static const List<RtcCapabilityCatalogEntry> entries = <RtcCapabilityCatalogEntry>[
    RtcCapabilityCatalogEntry(
      capabilityKey: "session",
      category: "required-baseline",
      surface: "cross-surface",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "join",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "publish",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "subscribe",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "mute",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "basic-events",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "health",
      category: "required-baseline",
      surface: "control-plane",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "unwrap",
      category: "required-baseline",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "screen-share",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "recording",
      category: "optional-advanced",
      surface: "control-plane",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "cloud-mix",
      category: "optional-advanced",
      surface: "control-plane",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "cdn-relay",
      category: "optional-advanced",
      surface: "control-plane",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "data-channel",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "transcription",
      category: "optional-advanced",
      surface: "control-plane",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "beauty",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "spatial-audio",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    RtcCapabilityCatalogEntry(
      capabilityKey: "e2ee",
      category: "optional-advanced",
      surface: "runtime-bridge",
    )
  ];

  const RtcCapabilityCatalog._();
}

List<RtcCapabilityCatalogEntry> getRtcCapabilityCatalog() {
  return RtcCapabilityCatalog.entries;
}

RtcCapabilityCatalogEntry? getRtcCapabilityDescriptor(String capabilityKey) {
  for (final entry in RtcCapabilityCatalog.entries) {
    if (entry.capabilityKey == capabilityKey) {
      return entry;
    }
  }

  return null;
}
